File: backend/tier1/unit_checker.py
"""Tier 1: Unit Confusion Checker

Detects when values appear to be entered in wrong units:
- Weight: kg entered as lb (or vice versa)
- Height: cm entered as m (or vice versa)
"""
from typing import List, Dict
from tier1.zscore import get_weight_range, get_median_weight, get_median_height

# Conversion factors
KG_TO_LB = 2.20462
LB_TO_KG = 0.453592
CM_TO_M = 0.01
M_TO_CM = 100


def check_unit_confusion_weight(value_kg: float, age_months: float, gender: str) -> List[Dict]:
    """
    Check if a weight value might be in wrong units.
    Returns list of alerts if issues found.
    """
    alerts = []
    min_w, max_w = get_weight_range(age_months, gender, sd_low=-4, sd_high=4)

    # Check: value seems reasonable as-is?
    in_range = min_w <= value_kg <= max_w

    if not in_range:
        # Check if value makes sense as pounds converted to kg
        as_kg_from_lb = value_kg * LB_TO_KG
        in_range_as_lb = min_w <= as_kg_from_lb <= max_w

        if in_range_as_lb and value_kg > max_w * 0.8:
            alerts.append({
                "alert_type": "UNIT_CONFUSION",
                "severity": "critical",
                "message": (
                    f"Weight {value_kg} kg is implausible for age {age_months:.0f}m. "
                    f"If interpreted as {value_kg} lb → {as_kg_from_lb:.1f} kg, "
                    f"it falls within normal range. Possible lb→kg confusion."
                ),
                "suggested_value": round(as_kg_from_lb, 2),
                "suggested_unit": "kg (converted from lb)",
            })

        # Check if value is implausibly small (kg entered as half?)
        as_kg_doubled = value_kg * KG_TO_LB
        in_range_doubled = min_w <= as_kg_doubled <= max_w

        if in_range_doubled and value_kg < min_w * 1.2:
            alerts.append({
                "alert_type": "UNIT_CONFUSION",
                "severity": "critical",
                "message": (
                    f"Weight {value_kg} kg is unusually low for age {age_months:.0f}m. "
                    f"If interpreted as kg→lb conversion error, actual weight might be "
                    f"~{as_kg_doubled:.1f} kg. Please verify."
                ),
                "suggested_value": round(as_kg_doubled, 2),
                "suggested_unit": "kg (possible reverse conversion)",
            })

    return alerts


def check_unit_confusion_height(value_cm: float, age_months: float, gender: str) -> List[Dict]:
    """
    Check if a height value might be in wrong units.
    Returns list of alerts if issues found.
    """
    alerts = []
    median_h = get_median_height(age_months, gender)
    min_h = median_h * 0.7
    max_h = median_h * 1.3

    in_range = min_h <= value_cm <= max_h

    if not in_range:
        # Check if value is in meters instead of cm
        if value_cm < 3.0:
            as_cm = value_cm * M_TO_CM
            if min_h <= as_cm <= max_h:
                alerts.append({
                    "alert_type": "UNIT_CONFUSION",
                    "severity": "critical",
                    "message": (
                        f"Height {value_cm} cm is implausibly low. "
                        f"If interpreted as {value_cm} m → {as_cm:.1f} cm, "
                        f"it falls within normal range. Possible m entered as cm."
                    ),
                    "suggested_value": round(as_cm, 1),
                    "suggested_unit": "cm (converted from m)",
                })

        # Check if cm entered as mm or vice versa
        if value_cm > max_h:
            as_cm_from_mm = value_cm / 10.0
            if min_h <= as_cm_from_mm <= max_h:
                alerts.append({
                    "alert_type": "UNIT_CONFUSION",
                    "severity": "high",
                    "message": (
                        f"Height {value_cm} cm is unusually tall. "
                        f"If interpreted as mm → {as_cm_from_mm:.1f} cm, "
                        f"it falls within normal range. Possible mm→cm error."
                    ),
                    "suggested_value": round(as_cm_from_mm, 1),
                    "suggested_unit": "cm (converted from mm)",
                })

    return alerts

File: backend/tier1/unit-checker.js
/**
 * Tier 1: Unit Checker
 * R001: Possible kg↔lb confusion
 * R002: Possible cm↔m or cm↔mm confusion
 */
const config = require('../config');

function checkUnitConfusion(observation, patient) {
  const alerts = [];

  if (observation.type === 'weight') {
    // R001: Check if value makes more sense in pounds
    const valueAsKg = observation.value;

    // If the kg value seems too high for age, maybe it's actually in pounds
    if (observation.unit === 'kg') {
      // A weight in pounds entered as kg would be ~2.2x too high
      const expectedMax = getExpectedWeightRange(patient).max;

      if (valueAsKg > expectedMax * 1.8 && valueAsKg * config.LB_TO_KG <= expectedMax) {
        alerts.push({
          alert_type: 'R001_UNIT_KG_LB',
          severity: config.SEVERITY.HIGH,
          message: `Weight ${valueAsKg} kg seems too high. Possible lb→kg confusion? ${valueAsKg} lb ≈ ${(valueAsKg * config.LB_TO_KG).toFixed(1)} kg`,
          explanation: `The recorded weight of ${valueAsKg} kg exceeds expected range. If this value is actually in pounds, the converted weight of ${(valueAsKg * config.LB_TO_KG).toFixed(1)} kg would be within normal range.`,
          suggested_value: Math.round(valueAsKg * config.LB_TO_KG * 10) / 10
        });
      }
    }
  }

  if (observation.type === 'height') {
    const value = observation.value;

    // R002: cm↔m confusion
    if (observation.unit === 'cm') {
      // If entered in meters but recorded as cm (e.g., 1.2 cm instead of 120 cm)
      if (value < 10 && value * config.M_TO_CM >= config.HEIGHT_MIN_CM) {
        alerts.push({
          alert_type: 'R002_UNIT_CM_M',
          severity: config.SEVERITY.HIGH,
          message: `Height ${value} cm seems too low. Possible m→cm confusion? ${value} m = ${value * config.M_TO_CM} cm`,
          explanation: `Height of ${value} cm is implausibly low. If this value is actually in meters, ${value * config.M_TO_CM} cm would be within normal range.`,
          suggested_value: value * config.M_TO_CM
        });
      }

      // If entered in mm but recorded as cm
      if (value > config.HEIGHT_MAX_CM && value / 10 >= config.HEIGHT_MIN_CM && value / 10 <= config.HEIGHT_MAX_CM) {
        alerts.push({
          alert_type: 'R002_UNIT_CM_MM',
          severity: config.SEVERITY.WARNING,
          message: `Height ${value} cm seems too high. Possible mm→cm confusion? ${value} mm = ${value / 10} cm`,
          explanation: `Height of ${value} cm exceeds physiological limits. If this value is in mm, ${value / 10} cm would be within normal range.`,
          suggested_value: value / 10
        });
      }
    }
  }

  return alerts;
}

function getExpectedWeightRange(patient) {
  const { ageInMonths } = require('./zscore');
  const months = ageInMonths(patient.date_of_birth);

  // Rough expected weight range by age (kg)
  const ranges = [
    { maxAge: 1, min: 2.5, max: 5 },
    { maxAge: 6, min: 4, max: 10 },
    { maxAge: 12, min: 6, max: 13 },
    { maxAge: 24, min: 8, max: 18 },
    { maxAge: 60, min: 10, max: 30 },
    { maxAge: 120, min: 15, max: 55 },
    { maxAge: 168, min: 25, max: 80 },
    { maxAge: 216, min: 35, max: 120 },
    { maxAge: Infinity, min: 40, max: 150 }
  ];

  for (const r of ranges) {
    if (months <= r.maxAge) return r;
  }
  return ranges[ranges.length - 1];
}

module.exports = { checkUnitConfusion };
